projects/RabbitMQ.Client/client/impl/Connection.cs

```diff
@@ -394,9 +397,93 @@ internal void OnCallbackException(CallbackExceptionEventArgs args)
             _callbackExceptionWrapper.Invoke(this, args);
         }
 
-        internal void Write(ReadOnlyMemory<byte> memory)
+        internal void Write(ReadOnlySpan<byte> payload)
+        {
+            if (!_frameHandler.IsClosed)
+            {
+                _writeSemaphore.Wait();
+                WriteSpan(payload);
+                _frameHandler.FrameWriter.FlushAsync().AsTask().Wait(); // Sync-over-async, not great, not terrible. All we can do for the sync path.
+                _writeSemaphore.Release();
+            }
+        }
+
+        internal async ValueTask WriteAsync(ReadOnlyMemory<byte> payload)
+        {
+            if (!_frameHandler.IsClosed)
+            {
+                if (!_writeSemaphore.Wait(0))
+                {
+                    await _writeSemaphore.WaitAsync().ConfigureAwait(false);
+                }
+
+                WriteSpan(payload.Span);
+                await _frameHandler.FrameWriter.FlushAsync().ConfigureAwait(false);
+                _writeSemaphore.Release();
```

Comment:
Are you deliberately not doing this in a finally block?

Reply:
Nope. I'll fix that